exercises/practice/linked-list/.approaches/circular-doubly-linked-list/content.md

```diff
@@ -0,0 +1,302 @@
+}
+
+static node_t *create_node(ll_data_t data, struct list *list)
+{
+   node_t *new_node = (node_t *)malloc(sizeof(node_t));
+   if (!new_node)
+      return NULL;
+   new_node->data = data;
+   if (list->head != NULL) {
+      new_node->prev = list->tail;
+      new_node->next = list->head;
+      list->tail->next = new_node;
+      list->head->prev = new_node;
+   } else {
+      new_node->prev = new_node;
+      new_node->next = new_node;
+      list->head = new_node;
+      list->tail = new_node;
+   }
+   list->length++;
+   return new_node;
+}
+
+static ll_data_t destroy_node(node_t *node, struct list *list)
+{
+   ll_data_t data = node->data;
+   node->prev->next = node->next;
+   node->next->prev = node->prev;
+   if (node == list->head)
+      list->head = node->next;
+   if (node == list->tail)
+      list->tail = node->prev;
+   node->prev = NULL;
+   node->next = NULL;
+   free(node);
+   node = NULL;
+   list->length--;
+   if (list->length == 0) {
+      list->head = NULL;
+      list->tail = NULL;
+   }
+   return data;
+}
+
+void list_push(struct list *list, ll_data_t item_data)
+{
+   list->tail = create_node(item_data, list);
+}
+
+ll_data_t list_pop(struct list *list)
+{
+   return destroy_node(list->tail, list);
+}
+
+void list_unshift(struct list *list, ll_data_t item_data)
+{
+   list->head = create_node(item_data, list);
+}
+
+ll_data_t list_shift(struct list *list)
+{
+   return destroy_node(list->head, list);
+}
+
+void list_delete(struct list *list, ll_data_t data)
+{
+   node_t *node = list->head;
+   while (node) {
+      if (node->data == data) {
+         destroy_node(node, list);
+         break;
+      }
+      if (node->next == list->head)
+         break;
+      node = node->next;
+   }
+}
+
+void list_destroy(struct list *list)
+{
+   node_t *node = list->head;
+   while (node) {
+      node_t *next_node = node->next;
+      destroy_node(node, list);
+      if (node == next_node)
+         break;
+      node = next_node;
+   }
+   free(list);
+   list = NULL;
+}
+```
+
+There are several ways of implementing a [circular doubly-linked list][circular-doubly-linked-list].
+This approach adds a `length` field which is initialized to `0` on list creation and updated as nodes are added and removed.
+
+A circular doubly-linked list populated with at least one node does not have `NULL` values for any node's `prev` or `next` fields.
+
+- If there is only one node in the list, its `prev` and `next` fields will point to itself.
+- If there are two nodes in a list, node `A` will point to node `B` for both its `prev` and `next` fields.
+Likewise, node `B` will point to node `A` for both its `prev` and `next` fields.
+- If there are three nodes in a list, node `A`'s `prev` will point at node `C` and node `A`'s `next` will point at node `B`.
+Node `B`'s `prev` will point at node `A` and node `B`'s `next` will point at node `C`.
+Node `C`'s `prev` will point at node `B` and node `C`'s `next` will point at node `A`.
```

Comment:
This explanation is good!
But the concept may be confusing for people not familiar with linked lists in general... It seems like an image would help; does exercism render graphs or images within markdown?